### ISSUE TITLE
Build read_audio function

### ISSUE THREAD

**Issue by liesenf** [elpaco-escience/scikit-talk#23]
**Objectives:**

1. build read_audio function
  - build module with read_audio function that reads .wav and .mp3
  - @n400peanuts discussion follow up: address memory leak by extracting audio without opening audio file, e.g. using base ffmpeg? 
2. load and save to json instead of csv
  - edit notebook code (that works with .csv data) to load json corpora instead
  - save extracted audio as audio.json instead of csv column

**Workflow**
- edit code `notebooks/Make_audio_column.ipynb` and build feature in `read_audio.py`
- required to merge #22

**Comment by n400peanuts** [comment]
@liesenf I am looking at the `get_row_audio` function. Is that the function you wish me to help you with? 

At the moment, this is what it is doing:

From looking at your code, it looks like that this function processes an audio file and extracts specific segments of audio based on the start and end times provided in the input DataFrame df_audio. The extracted audio segments and the sampling rate are added to the DataFrame as new columns.

Thus It returns a "modified" `df_audio` DataFrame containing the audio segments and their sampling rate. Is that what you wish to have? a pandas dataframe containing the start, end of segments and sampling rate? 

For some reasons I remembered you wanted to _create_ new files from the start/end info.

**Comment by liesenf** [comment]
Yes, I used `get_row_audio` to add the `audio` and `rate` column to a data frame. 
- since we are moving away from data frames, the function will eventually have to be rewritten with json as input and output but that can be done later.
- the other issue is exploring the use of `ffmpeg` or some other tool that doesn't require reading the whole audio file (what we talked about). Can you have a look at that?

**Comment by n400peanuts** [comment]
Ok, I'll take a look now.

**Comment by liesenf** [comment]
Great! If needed,

here is a test data frame with some timestamps (on the `corpusbuilder` branch):
[https://github.com/elpaco-escience/scikit-talk/blob/csv_to_json/data/ulwa_testdata_sktalk_format.csv](https://github.com/elpaco-escience/scikit-talk/blob/csv_to_json/data/ulwa_testdata_sktalk_format.csv)


and here audio data for that corpus (on pony land): 

`smb://tensusers2-srv.science.ru.nl/tensusers2/aliesenfeld/Elpaco dataset/Elpaco dataset/ulwa1/`

**Comment by n400peanuts** [comment]
Fantastic, thanks!!!!


Il giorno 24 apr 2023, alle ore 15:34, Andreas Liesenfeld ***@***.***> ha scritto:

﻿

Great! If needed,

here is a test data frame with some timestamps (on the corpusbuilder branch):
https://github.com/elpaco-escience/scikit-talk/blob/csv_to_json/data/ulwa_testdata_sktalk_format.csv

and here audio data for that corpus (on pony land):

[smb://tensusers2-srv.science.ru.nl/tensusers2/aliesenfeld/Elpaco dataset/Elpaco dataset/ulwa1/](smb://tensusers2-srv.science.ru.nl/tensusers2/aliesenfeld/Elpaco dataset/Elpaco dataset/ulwa1/)

**Comment by n400peanuts** [comment]
@liesenf after some [stackoverflow](https://stackoverflow.com/questions/7465153/command-to-get-information-about-a-mp3-using-ffmpeg) and digging into the [ffmpeg](https://ffmpeg.org/ffprobe.html) page, I made a function for you to show how to use ffmpeg/ffprobe to get the sampling rate without opening the audio file. It basically relies on the `subprocess` module to run in the terminal the commands stored in 'cmd'. 
I think you can add it into the `get_row_audio` function. Let me know how it goes.


``` 
import subprocess
import json

def get_sampling_rate(file_path):
    cmd = [
        "ffprobe",
        "-v", "quiet",
        "-print_format", "json",
        "-show_streams",
        file_path
    ]

    result = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
    output = json.loads(result.stdout)

    for stream in output["streams"]:
        if stream["codec_type"] == "audio":
            return int(stream["sample_rate"])

    raise ValueError("No audio stream found in the file")

# Replace 'list_audio_3_40_balanced.wav' with the path to your audio file
file_path = 'list_audio_3_40_balanced.wav'

sampling_rate = get_sampling_rate(file_path)
print(sampling_rate)
```

NB: you must have installed ffmpeg in your system though
